src/realm/alloc.hpp

```diff
@@ -35,6 +35,10 @@ class Replication;
 
 using ref_type = size_t;
 
+// Maximum .realm file size and also maximum in-memory transaction snapshot size. On 32-bit architectures we need 
+// to cap it to 2^31 * 0.80 bytes because arithmetic on references can otherwise overflow in various places.
+constexpr size_t maximum_realm_size = static_cast<size_t>((std::numeric_limits<ref_type>::max() >> 1) * 0.8);
```

Comment:
@ironage: I agree with the first observation - we were moving a bit too fast yesterday. We'll shave the goat closer. What do you mean by "where the size multiplies" ?